### PR TITLE
Add getRecordTypeId method to fflib_SObjectDescribe.cls

### PR DESCRIPTION
**Pull request by PeterLinBiz**
Changes to be committed:
	modified:   sfdx-source/apex-common/main/classes/fflib_SObjectDescribe.cls
	modified:   sfdx-source/apex-common/test/classes/fflib_SObjectDescribeTest.cls

---
This change is [Reviewable](https://reviewable.io/reviews/apex-enterprise-patterns/fflib-apex-common/363)

### CODE REVIEW

**Comment by PeterLinBiz** [comment]
Dear AEP team, in one org, I have more than one hundred calls to the getRecordTypeId method.

**Review by stohn777 — CHANGES_REQUESTED** [review]
Hi @peterlinbiz.  Thanks for the PR!  Just a couple of minor point included below.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/apex-enterprise-patterns/fflib-apex-common/363#-:-MjzTU9I-WlC8_vY4XaY:b-t7v7ri)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @PeterLinBiz)

___
*[sfdx-source/apex-common/test/classes/fflib_SObjectDescribeTest.cls, line 184 at r1](https://reviewable.io/reviews/apex-enterprise-patterns/fflib-apex-common/363#-MjzSwgd6C5Cssi96Fzg:-MjzSwgd6C5Cssi96Fzh:b-klgd81) ([raw file](https://github.com/apex-enterprise-patterns/fflib-apex-common/blob/77e46d3c11f6d807bf9a8156a7f9195acc42ca7a/sfdx-source/apex-common/test/classes/fflib_SObjectDescribeTest.cls#L184)):*
> ```VBA
> 
> 	@IsTest
> 	static void testGetRecordTypeId(){
> ```

For consistency with other test names, change to "getRecordTypeId".
___
*[sfdx-source/apex-common/test/classes/fflib_SObjectDescribeTest.cls, line 188 at r1](https://reviewable.io/reviews/apex-enterprise-patterns/fflib-apex-common/363#-MjzTHFm1H99M_n4VyV0:-MjzTHFm1H99M_n4VyV1:b-r2dczy) ([raw file](https://github.com/apex-enterprise-patterns/fflib-apex-common/blob/77e46d3c11f6d807bf9a8156a7f9195acc42ca7a/sfdx-source/apex-common/test/classes/fflib_SObjectDescribeTest.cls#L188)):*
> ```VBA
> 		Id idStringInvalid = fflib_SObjectDescribe.getRecordTypeId('Account', 'InvalidInvalid');
> 		Id idToken = fflib_SObjectDescribe.getRecordTypeId(Account.SObjectType, 'GUPA');
> 		Id idTokenInvlid = fflib_SObjectDescribe.getRecordTypeId(Account.SObjectType, 'InvalidInvalid');
> ```

Misspelling.  Probably should be "idTokenInvalid"
___
*[sfdx-source/apex-common/test/classes/fflib_SObjectDescribeTest.cls, line 200 at r1](https://reviewable.io/reviews/apex-enterprise-patterns/fflib-apex-common/363#-MjzT67a9woaYWDAX1Cv:-MjzT67a9woaYWDAX1Cw:b1ddfxr) ([raw file](https://github.com/apex-enterprise-patterns/fflib-apex-common/blob/77e46d3c11f6d807bf9a8156a7f9195acc42ca7a/sfdx-source/apex-common/test/classes/fflib_SObjectDescribeTest.cls#L200)):*
> ```VBA
> 			System.assertEquals(idString, listRecordTypes[0].Id);
> 			System.assertEquals(idToken, listRecordTypes[0].Id);
> 		} else {
> ```

I don't believe this else-block is adding value given the if condition.

**Comment by PeterLinBiz** [comment]
Hi @stohn777 Thanks for your review!
I have removed the else-block and modified the other two lines.

**Review by PeterLinBiz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/apex-enterprise-patterns/fflib-apex-common/363)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @stohn777)

___
*[sfdx-source/apex-common/test/classes/fflib_SObjectDescribeTest.cls, line 184 at r1](https://reviewable.io/reviews/apex-enterprise-patterns/fflib-apex-common/363#-MjzSwgd6C5Cssi96Fzg:-Mk07C7Q6IWnmTEmKrUP:bet9uz8) ([raw file](https://github.com/apex-enterprise-patterns/fflib-apex-common/blob/77e46d3c11f6d807bf9a8156a7f9195acc42ca7a/sfdx-source/apex-common/test/classes/fflib_SObjectDescribeTest.cls#L184)):*
<details><summary><i>Previously, stohn777 (John Storey) wrote…</i></summary><blockquote>

For consistency with other test names, change to "getRecordTypeId".
</blockquote></details>

I have modified this in the new commit.
___
*[sfdx-source/apex-common/test/classes/fflib_SObjectDescribeTest.cls, line 188 at r1](https://reviewable.io/reviews/apex-enterprise-patterns/fflib-apex-common/363#-MjzTHFm1H99M_n4VyV0:-Mk07KOt7W0Tb3hAHPpC:blo8gjs) ([raw file](https://github.com/apex-enterprise-patterns/fflib-apex-common/blob/77e46d3c11f6d807bf9a8156a7f9195acc42ca7a/sfdx-source/apex-common/test/classes/fflib_SObjectDescribeTest.cls#L188)):*
<details><summary><i>Previously, stohn777 (John Storey) wrote…</i></summary><blockquote>

Misspelling.  Probably should be "idTokenInvalid"
</blockquote></details>

I have modified this to "idTokenInvalid" in the new commit.
___
*[sfdx-source/apex-common/test/classes/fflib_SObjectDescribeTest.cls, line 200 at r1](https://reviewable.io/reviews/apex-enterprise-patterns/fflib-apex-common/363#-MjzT67a9woaYWDAX1Cv:-Mk07WpJ989pVoQbA5sF:b-f7xms6) ([raw file](https://github.com/apex-enterprise-patterns/fflib-apex-common/blob/77e46d3c11f6d807bf9a8156a7f9195acc42ca7a/sfdx-source/apex-common/test/classes/fflib_SObjectDescribeTest.cls#L200)):*
<details><summary><i>Previously, stohn777 (John Storey) wrote…</i></summary><blockquote>

I don't believe this else-block is adding value given the if condition.
</blockquote></details>

I have removed the else-block.

File: sfdx-source/apex-common/main/classes/fflib_SObjectDescribe.cls [hunk 1]
```diff
@@ -216,6 +216,19 @@ public class fflib_SObjectDescribe {
 		return getDescribe(instance.getSObjectType());
 	}
 
+	private Id getRecordTypeId(String recordTypeDeveloperName)
+	{
+		return getDescribe().getRecordTypeInfosByDeveloperName().get(recordTypeDeveloperName)?.getRecordTypeId();
+	}
+
+	public static Id getRecordTypeId(String sObjectName, String recordTypeDeveloperName){
```
Comment (wimvelzeboer) [inline sfdx-source/apex-common/main/classes/fflib_SObjectDescribe.cls:224]: Small detail; 
I see an inconsistency in brackets placements, 
opening brackets are usually places on a new line (like line 220)

Comment (PeterLinBiz) [reply sfdx-source/apex-common/main/classes/fflib_SObjectDescribe.cls:224]: I am changing this in the next commit now.
Thanks @wimvelzeboer !

Comment (PeterLinBiz) [reply sfdx-source/apex-common/main/classes/fflib_SObjectDescribe.cls:223]: I have completed the change in the pushed commit.

**Review by stohn777 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/apex-enterprise-patterns/fflib-apex-common/363)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @ImJohnMDaniel, @PeterLinBiz, and @wimvelzeboer)

*[a discussion](https://reviewable.io/reviews/apex-enterprise-patterns/fflib-apex-common/363#-Mk2fIqt8YdUmpYI5udT:-Mk2pEP4BP1eAZDrMJdg:b37dzsv) (no related file):*
<details><summary><i>Previously, daveespo (David Esposito) wrote…</i></summary><blockquote>

@peterlinbiz  -- can you say more about why you think this PR is a necessary addition to the fflib_SObjectDescribe class? There are a million methods on DescribeSObjectResult -- why do the Record Type getters deserve a place in the public interface for fflib_SObjectDescribe?

Second, all of the other static methods on fflib_SObjectDescribe are for getting an instance of fflib_SObjectDescribe -- so I don't think that these methods should be static; they should be instance methods

@imjohnmdaniel and @stohn777  -- before Peter reworks the solution, can you weigh in on whether this should be a PR at all?
</blockquote></details>

Upon reviewing this new feature, I first looked for existing RecordType functionality, as defined here, and found none.  With RecordType being a subset of SObjectDescribe, I was not alarmed by the change.  However with further concern raised, I can see a couple of obvious paths.  One, is to let the RecordType functionality stand as written here, or two, move it to a class that's more specific, perhaps something like "fflib_RecordType".

Beyond just the maintainers perspectives, I'd like to hear from the community, too, about the proper manner through which to deliver this functionality.

---

**Comment by PeterLinBiz** [comment]
@daveespo Thank you for your very good questions.

First,
I don't have persuasive argument about how important this method "fflib_IDGenerator.generate(Account.SObjectType)" is comparing to other probably more important non-@IsTest method in fflib.
I can only say that in one of my production org, I used this method for more than a hundred times for the convenience of my test class.
If I didn't see the above method in fflib, I would not know how long it will take for me to figure out this convenient way to generate an Id for testing purposes.
So, "fflib_IDGenerator.generate(Account.SObjectType)" has been making my life much easier as a developer. And knowing and using this method have been an enjoyable learning life journey for me.

Similarly, in one of my production org, we see more than a hundred lines of diverse ways to get the RecordTypeId as below.
1. SOQL
2. Account.SObjectType.getDescribe().getRecordTypeInfosByDeveloperName().get(recordTypeDeveloperName)?.getRecordTypeId();
3. HomeGrownClass.getRecordTypeId('Account', recordTypeDeveloperName);

So, if "fflib_SObjectDescribe.getRecordTypeId(Schema.SObjectType token, String recordTypeDeveloperName)" may make the life of a developer easier similar to "fflib_IDGenerator.generate(Account.SObjectType)", you may want to consider adding it.

Second, regarding static or instance methods, let's compare them for the purposes of making a developer's life easier.
1. Static method as now in this PR.
Getting a RecordTypeId is as below.
```java
Id idRecordTypeOne = fflib_SObjectDescribe.getRecordTypeId(Schema.SObjectType token, String recordTypeDeveloperName);
Id idRecordTypeTwo = fflib_SObjectDescribe.getRecordTypeId(String sObjectName, String recordTypeDeveloperName);
```
2. Instance method which is not in this PR.
(To implement this, I would change line 219 as public and delete line 222 through 230 which is a simpler implementation.)
```java
Id idRecordTypeOne 
        = fflib_SObjectDescribe.getDescribe(Schema.SObjectType token).getRecordTypeId(String recordTypeDeveloperName);
Id idRecordTypeTwo
        = fflib_SObjectDescribe.getDescribe(String sObjectName).getRecordTypeId(String recordTypeDeveloperName);
```

So, for static method, the end users only need to call one public static overload method with two parameters, which IMO, making the life of the end users easier. By choosing static method, the end users don't need to know or to use the private getRecordTypeId instance method, however from the fflib point of view, we have extra codes to maintain which are the two public static overload methods comparing to the instance method option below.

For public instance method, the end users will need to be educated (if with less experience) to call two chained methods to get a RecordTypeId as below.  And the good side of from fflib point of view, we don't need lines between 222 and 230 in this PR.
```java
Id idRecordTypeOne = fflib_SObjectDescribe.getDescribe(Account.SObjectType).getRecordTypeId('GUPA');
Id idRecordTypeTwo = fflib_SObjectDescribe.getDescribe('Account').getRecordTypeId('GUPA');
```

Third,
fflib libraries are all great!
When more and more teams starting to implement fflib, there will always be challenging to balance (for the end users to create Unlocked Packages versions) easier to adapt and to adopt comparing to easier (for the AEP team) to maintain and to keep it reliable and robust.

During the first few months of my own experience after installed the Unlocked Pakcages of fflib common and mocks, I didn't realize the beauty of the private static method instanceCache in fflib_SObjectDescribe. If there was a getRecordTypeId method in fflib_SObjectDescribe back then, I would not need to develop and refactor a few more versions of my HomeGrownClass.getRecordTypeId('Account', recordTypeDeveloperName) which contains my own way of calling something similar to "Account.SObjectType.getDescribe().getRecordTypeInfosByDeveloperName()".

If this PR is not accepted, I will probably modify my HomeGrownClass as below to take the advantage of the **beauty** of the **instanceCache** method in fflib_SObjectDescribe.

```java
public class HomeGrownClass{
    private Id getRecordTypeId(DescribeSObjectResult describeSObjectResult, String recordTypeDeveloperName){
        return describeSObjectResult.getRecordTypeInfosByDeveloperName().get(recordTypeDeveloperName)?.getRecordTypeId();
    }

    public static Id getRecordTypeId(String sObjectName, String recordTypeDeveloperName){
        Schema.DescribeSObjectResult describeSObjectResult = fflib_SObjectDescribe.getDescribe(sObjectName).getDescribe();
        return new HomeGrownClass().getRecordTypeId(describeSObjectResult, recordTypeDeveloperName);
    }

    public static Id getRecordTypeId(Schema.SObjectType token, String recordTypeDeveloperName){
        Schema.DescribeSObjectResult describeSObjectResult = fflib_SObjectDescribe.getDescribe(token).getDescribe();
        return new HomeGrownClass().getRecordTypeId(describeSObjectResult, recordTypeDeveloperName);
    }
}
```

And I do hope this "static with one getRecordTypeId with overload two parameters" (personally I like the token version which doesn't have to call the wrappedGlobalDescribe method which may save some memory and CPU time) version of this PR will be accepted so that this may make the developer's life a little bit easier with one method call similar to "fflib_IDGenerator.generate(Account.SObjectType)". (Because most of my calls to this HomeGrownClass.getRecordTypeId method is string version, I will need to plan to refactor them to the token version).

Thank you, AEP team for keeping this AEP as reliable and robust as possible!!

**Comment by PeterLinBiz** [comment]
> Previously, daveespo (David Esposito) wrote…
> Upon reviewing this new feature, I first looked for existing RecordType functionality, as defined here, and found none. With RecordType being a subset of SObjectDescribe, I was not alarmed by the change. However with further concern raised, I can see a couple of obvious paths. One, is to let the RecordType functionality stand as written here, or two, move it to a class that's more specific, perhaps something like "fflib_RecordType".
> 
> Beyond just the maintainers perspectives, I'd like to hear from the community, too, about the proper manner through which to deliver this functionality.

@stohn777
If we go with "fflib_RecordType" so that to add one public static overload method of "getId", it probably will look very clean similar to the HomeGrownClass example in my previous comment.

**Review by daveespo — COMMENTED** [review]
@peterlinbiz  - We discussed this PR again today and decided to reject it. `fflib_SObjectDescribe` is mostly an internal implementation detail for supporting fflib_QueryFactory

While it's admirable that you're using it standalone, we are going to reject PRs that are more 'utility' in nature. It seems like the functionality you're requesting could and should be part of its own library. 

> *[Reviewable](https://reviewable.io/reviews/apex-enterprise-patterns/fflib-apex-common/363#-:-MzuQX9P7C6jCbTrTyKw:b-k0no82)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @daveespo, @ImJohnMDaniel, @PeterLinBiz, @stohn777, and @wimvelzeboer)

**Comment by wimvelzeboer** [comment]
@daveespo I agree with that this is something that should be separate from `fflib_SObjectDescribe`. @PeterLinBiz please have a look at [this implementation](https://github.com/wimvelzeboer/fflib-apex-extensions/blob/main/sfdx-source/apex-extensions/main/default/classes/utils/fflib_RecordTypeId.cls) that I made in [fflib-apex-extensions](https://github.com/wimvelzeboer/fflib-apex-extensions), where it is implemented as a Utility class.

**Pull request closed without merging**